modules/dnn/src/onnx/onnx_importer.cpp

```diff
@@ -1759,8 +1766,27 @@ void ONNXImporter::parseLSTM(LayerParams& layerParams, const opencv_onnx::NodePr
     bool need_y = lstm_proto.output_size() > 0 && !lstm_proto.output(0).empty();
 
     const std::string y_name = need_y ? lstm_proto.output(0) : "";
-    const std::string yh_name = need_yh ? lstm_proto.output(1) : "";
-    const std::string yc_name = need_yc ? lstm_proto.output(2) : "";
+    const std::string yh_name = [&](){
+        if(constBlobs.find(lstm_proto.input(5)) == constBlobs.end() &&
+           constBlobs.find(lstm_proto.input(6)) == constBlobs.end() && 
+           num_directions == 1)
+           {
+                return need_yh ? lstm_proto.output(2) : "";
```

Comment:
This looks counterintuitive for the user's perspective. Why does the number of dynamic inputs change the order of outputs? I think it should be addressed in the lstm::forward method.

Reply:
Forward method is working as expected (output[0] contains hidden state and output[1] contains cell state) regardless of number if dynamic inputs and their order. 

1. The problem though is that is LSTM layer returns `cell_state` instead of `hidden state` for simple single layer onnx model as depicted in [#21118](https://github.com/opencv/opencv/issues/21118) 
2. For complex ONNX models with LSTM layers in it, solution that was suggested before this fix works just fine (test have covered such serious). However, the logic breaks down for single LSTM layer ONNX model. It's output is incorrect. 

The only feasible solution found was this. Since it does not break logic of other type of layers. 

If there is a better solution, could you please provide me a hint.

Reply:
Could we address p. 1? It seems like it would fix this as well.

Reply:
Removed dynamic order determination. 
Actually this unexpected behaviour was cause by 

`return self.lstm(x, (hx, cx))` 

statement in my test case instead of 

`return self.lstm(x, (hx, cx))[0]` 


as in the original test cases